core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala

```diff
@@ -417,8 +422,90 @@ class ColumnFormatRelation(
+    val indexTblName = snc.getIndexTable(indexIdent).toString
+    val caseInsensitiveMap = new CaseInsensitiveMutableHashMap(tableRelation.origOptions)
+    val tempOptions = caseInsensitiveMap.filterNot(pair => {
+      pair._1.equals(Utils.toLowerCase(StoreUtils.PARTITION_BY)) ||
+        pair._1.equals(Utils.toLowerCase(StoreUtils.COLOCATE_WITH)) ||
+        pair._1.equals(Utils.toLowerCase(JdbcExtendedUtils.DBTABLE_PROPERTY)) ||
+        pair._1.equals(Utils.toLowerCase(ExternalStoreUtils.INDEX_NAME))
+    }).toMap + (StoreUtils.PARTITION_BY -> indexColumns.mkString(",")) +
+      (StoreUtils.GEM_INDEXED_TABLE -> tableIdent.toString) +
+      (JdbcExtendedUtils.DBTABLE_PROPERTY -> indexTblName)
+
+    val indexOptions = parameters.get(StoreUtils.COLOCATE_WITH) match {
+      case Some(value) => tempOptions + (StoreUtils.COLOCATE_WITH -> value)
+      case _ => tempOptions
+    }
+    snc.createTable(
+      indexTblName,
+      "column",
+      tableRelation.schema,
+      indexOptions)
+  }
+
+  override def createIndex(indexIdent: QualifiedTableName,
+                           tableIdent: QualifiedTableName,
+                           indexColumns: Seq[String],
+                           options: Map[String, String]): Unit = {
+
+    createIndexTable(indexIdent, tableIdent, this, indexColumns, options)
+  }
+
+  override def addDependent(dependent: DependentRelation,
+                            catalog: SnappyStoreHiveCatalog): Boolean =
+    DependencyCatalog.addDependent(table, dependent.name)
+
+  override def removeDependent(dependent: DependentRelation,
+                               catalog: SnappyStoreHiveCatalog): Boolean =
+    DependencyCatalog.removeDependent(table, dependent.name)
+
+  override def getDependents(catalog: SnappyStoreHiveCatalog): Seq[String] =
+    DependencyCatalog.getDependents(table)
 }
 
+/**
+  * Currently this is same as ColumnFormatRelation but has kept it as a separate class
+  * to allow adding of any index specific functionality in future.
+  */
+class IndexColumnFormatRelation(
+                            _table: String,
+                            _provider: String,
+                            _mode: SaveMode,
+                            _userSchema: StructType,
+                            _schemaExtensions: String,
+                            _ddlExtensionForShadowTable: String,
+                            _origOptions: Map[String, String],
+                            _externalStore: ExternalStore,
+                            _blockMap: Map[InternalDistributedMember, BlockManagerId],
+                            _partitioningColumns: Seq[String],
+                            _context: SQLContext,
+                            baseTableName: String)
+  extends ColumnFormatRelation(_table, _provider, _mode, _userSchema,
+    _schemaExtensions, _ddlExtensionForShadowTable, _origOptions,
+    _externalStore, _blockMap, _partitioningColumns, _context)
+ with DependentRelation {
+
```

Comment:
This is a bit of a problem since ColumnFormatRelation is a ParentRelation. So Index becomes both parent and dependent? Perhaps we should stick trait in ColumnFormatRelation elsewhere when creating, or split into base class.